docs/notebooks/Hidden_Linear_Function_Problem.ipynb

```diff
@@ -0,0 +1,527 @@
+    "\n",
+    "Also define\n",
+    "\n",
+    "$$\\mathcal{L}_q = \\Big\\{x \\in  \\mathbb{F}_2^n : q(x \\oplus y) = (q(x) + q(y)) ~\\text{mod}~ 4 ~~ \\forall y \\in \\mathbb{F}_2^n \\Big\\}.$$\n",
+    "\n",
+    "Turns out (see Lemma 1 on page 6 in [1]) that restriction of $q$ on $\\mathcal{L}_q$ is a linear function, i.e. there exists such $z \\in \\mathbb{F}_2^n$, that\n",
+    "\n",
+    "$$q(x) = 2 z^T x ~~\\forall x \\in \\mathcal{L}_q.$$\n",
+    "\n",
+    "Our task is, given $A$ and $b$, to find $z$. There may be multiple answers - we need to find any such answer.\n",
+    "\n",
+    "### Why is this problem interesting?\n",
+    "\n",
+    "#### 1. It's a problem without an oracle\n",
+    "\n",
+    "This problem is similar to a problem solved by [Bernstein–Vazirani algorithm](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Vazirani_algorithm). It also finds coefficients of unknown linear function. But in the Bernstein-Vazirani algorithm this function is represented by an oracle. In this problem, the linear function is \"hidden\" in inputs $A$ and $b$.\n",
+    "\n",
+    "#### 2. Quantum circuits have advantage over classical when solving this problem\n",
+    "\n",
+    "As we will show below, this problem can be solved with a Clifford circuit. Therefore, according to the [Gottesman–Knill theorem](https://en.wikipedia.org/wiki/Gottesman%E2%80%93Knill_theorem), this problem can be solved in polynomial time on a classical computer. So, it might look like quantum computers aren't better than classical ones in solving this problem.\n",
+    "\n",
+    "However, if we apply certain restrictions on matrix $A$, the circuit will have fixed depth (i.e. number of Moments). Namely, if the matrix $A$ is an adjacency matrix of a \"grid\" graph (whose edges can be colored in 4 colors), all CZ gates will fit in 4 moments, and overall we will have only 8 moments - and this doesn't depend on $n$.\n",
+    "\n",
+    "But for classical circuits it can be proven (see [1]) that even if we restrict matrix $A$ in the same way, the depth of classical circuit (with gates of bounded fan-in) must grow as $n$ grows (in fact, it grows as $\\log(n)$).\n",
+    "\n",
+    "### Structure of this notebook\n",
+    "\n",
+    "In the first part, we solve this problem in a straightforward way, which follows directly from the problem definition. We will use this solution to verify our quantum solution in the next part. Also this part contains helper code to generate \"interesting\" instances of the problem.\n",
+    "\n",
+    "In the next part we solve this problem with Cirq. First, we write a code which builds a quantum circuit for solving arbitrary instances of the problem. Then we use Cirq's Clifford simulator to simulate this circuit. We do it for small instances and compare results to the bruteforce solution from the previous part. Then we solve the problem for a larger instance of the problem to demonstrate that it can be solved efficiently.\n",
```

Comment:
Should be a comma after "Then, ".  Brute force should be two words. Several instances of "bruteforce" elsewhere as well.